consensus/qbft/core/preprepare.go

```diff
@@ -0,0 +1,128 @@
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/consensus"
+	"github.com/ethereum/go-ethereum/consensus/qbft/message"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+func (c *core) sendPreprepareMsg(request *Request) {
+	logger := c.logger.New("state", c.state)
+
+	// If I'm the proposer and I have the same sequence with the proposal
+	if c.current.Sequence().Cmp(request.Proposal.Number()) == 0 && c.IsProposer() {
+		curView := c.currentView()
+		preprepare := message.NewPreprepare(curView.Sequence, curView.Round, request.Proposal)
+		preprepare.SetSource(c.Address())
+
+		// Sign payload
+		encodedPayload, err := preprepare.EncodePayloadForSigning()
+		if err != nil {
+			logger.Error("QBFT: Failed to encode payload of pre-prepare message", "msg", preprepare, "err", err)
+			return
+		}
+		signature, err := c.backend.Sign(encodedPayload)
+		if err != nil {
+			logger.Error("QBFT: Failed to sign pre-prepare message", "msg", preprepare, "err", err)
+			return
+		}
+		preprepare.SetSignature(signature)
+
+		// Justification
+		if request.RCMessages != nil {
+			preprepare.JustificationRoundChanges = make([]*message.SignedRoundChangePayload, 0)
+			for _, m := range request.RCMessages.Values() {
+				preprepare.JustificationRoundChanges = append(preprepare.JustificationRoundChanges, &m.(*message.RoundChange).SignedRoundChangePayload)
+				logger.Info("QBFT: Appending RC justification", "rc", m.(*message.RoundChange).SignedRoundChangePayload)
+			}
+			logger.Info("QBFT: On Pre-prepare", "rc justification", preprepare.JustificationRoundChanges)
+		}
+		if request.PrepareMessages != nil {
+			preprepare.JustificationPrepares = request.PrepareMessages
+			logger.Info("QBFT: On Pre-prepare", "prepare justification", preprepare.JustificationPrepares)
+		}
+
+		// RLP-encode message
+		payload, err := rlp.EncodeToBytes(&preprepare)
+		if err != nil {
+			logger.Error("QBFT: Failed to encode pre-prepare message", "msg", preprepare, "err", err)
+			return
+		}
+
+		logger.Info("QBFT: sendPreprepareMsg", "m", preprepare, "payload", hexutil.Encode(payload))
+		// Broadcast RLP-encoded message
+		if err = c.backend.Broadcast(c.valSet, preprepare.Code(), payload); err != nil {
+			logger.Error("QBFT: Failed to broadcast message", "msg", preprepare, "err", err)
+			return
+		}
+
+		// Set the preprepareSent to the current round
+		c.current.preprepareSent = curView.Round
+	}
+}
+
+func (c *core) handlePreprepareMsg(preprepare *message.Preprepare) error {
+	logger := c.logger.New("state", c.state)
+
+	c.logger.Info("QBFT: handlePreprepareMsg", "view", preprepare.View(), "m", preprepare)
+
+	// Check if the message comes from current proposer
+	logger.Warn("QBFT who's proposer?", "source", preprepare.Source(), "proposer", c.valSet.GetProposer().Address())
```

Comment:
If I understand correctly this will log a warning message on every received `preprepare` message.

Instead should we log source address on the above Info log "QBFT: handlePreprepareMsg" and Warn only if the proposer is invalid